fuse/pathfs/pathfs.go

```diff
@@ -255,147 +242,100 @@ func (n *pathInode) SetInode(node *nodefs.Inode) {
 	n.inode = node
 }
 
-// Reread all client nodes below this node.  Must run outside the treeLock.
-func (n *pathInode) updateClientInodes() {
-	n.GetAttr(&fuse.Attr{}, nil, nil)
-	for _, ch := range n.Inode().FsChildren() {
-		ch.Node().(*pathInode).updateClientInodes()
+// getParent - returns a random parent of this pathInode and the name under this
+// parent.
+// The caller must hold pathLock.
+func (n *pathInode) getParent() (*pathInode, string) {
+	for k, _ := range n.parents {
+		return k.parent, k.name
 	}
+	return nil, ""
 }
 
-// GetPath returns the path relative to the mount governing this
-// inode.  It returns nil for mount if the file was deleted or the
-// filesystem unmounted.
-func (n *pathInode) GetPath() string {
+// GetPath returns a path relative to the mount pointing to this pathInode.
+// Due to hard links, a file may have many paths. This function returns a
+// random one.
+// Must be called with pathLock held.
+func (n *pathInode) getPath() string {
 	if n == n.pathFs.root {
 		return ""
 	}
-
-	pathLen := 0
-
-	// The simple solution is to collect names, and reverse join
-	// them, them, but since this is a hot path, we take some
-	// effort to avoid allocations.
-
-	n.pathFs.pathLock.RLock()
-	p := n
-	for ; p.Parent != nil; p = p.Parent {
-		pathLen += len(p.Name) + 1
+	if len(n.parents) == 0 {
+		dummyPath := fmt.Sprintf(".ERROR_NO_PARENTS.%d", n.clientInode)
+		log.Printf("getPath: non-root node without parents, returning dummyPath=%q", dummyPath)
+		return dummyPath
 	}
-	pathLen--
 
-	if p != p.pathFs.root {
-		n.pathFs.pathLock.RUnlock()
-		return ".deleted"
+	// The simple solution is to collect names, and reverse join
+	// them. This is a hot path and may need optimization to avoid
+	// too many allocations.
+
+	// Walk up to the root and collect names in pathSegments
+	var pathSegments []string
+	var nextName string
+	nextParent := n
+	for len(nextParent.parents) > 0 {
+		nextParent, nextName = nextParent.getParent()
+		pathSegments = append(pathSegments, nextName)
+
+		if len(pathSegments) > syscall.PathMax {
```

Comment:
Yes it is, it helped during development, but if your filesystem does not
call into the kernel there is no real reason to limit the length. Will drop
the check.